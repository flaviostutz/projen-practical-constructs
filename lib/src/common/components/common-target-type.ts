export enum CommonTargets {
  PREPARE = 'prepare',
  BUILD = 'build',
  INSTALL = 'install',
  COMPILE = 'compile',
  PACKAGE = 'package',
  LINT = 'lint',
  LINT_FIX = 'lint-fix',
  TEST = 'test',
  RELEASE = 'release',
  DEPLOY = 'deploy',
  PUBLISH = 'publish',
}
