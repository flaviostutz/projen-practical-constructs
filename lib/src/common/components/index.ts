export * from './base-tooling';
export * from './common-target-type';
export * from './common-targets';
export * from './release-tasks';
export * from './publish';
