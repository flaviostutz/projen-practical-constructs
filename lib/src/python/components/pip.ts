/* eslint-disable no-new */
import { Component, DependencyType, Project, TaskRuntime } from 'projen';

import { TaskOptions } from '../tasks';

const PIP_TOOLS_VERSION = '==7.4.1';

export class Pip extends Component {
  constructor(project: Project, taskOpts: TaskOptions, opts?: PipOptions) {
    super(project);

    // default options
    const optsd: PipOptions = {
      pythonExec: 'python',
      lockFile: 'constraints.txt',
      lockFileDev: 'constraints-dev.txt',
      ...opts,
    };

    // add git ignore patterns
    project.addGitIgnore(`${taskOpts.venvPath}/`);
    project.addGitIgnore('*.egg-info');
    project.addGitIgnore('build');
    project.addGitIgnore('*.pyc');
    project.addGitIgnore('__pycache__');
    project.addGitIgnore('.cache');

    // add build related tasks
    project.tasks.addTask('install', {
      description: `Install dependencies from ${optsd.lockFile}`,
      exec: `${taskOpts.venvPath}/bin/pip install --require-virtualenv -c ${optsd.lockFile}`,
    });

    const installDevTask = project.tasks.addTask('install-dev', {
      description: `Install dependencies from ${optsd.lockFileDev}`,
      exec: `${taskOpts.venvPath}/bin/pip install --require-virtualenv -c ${optsd.lockFileDev} --editable .[dev]`,
    });

    const prepareVenvTask = project.tasks.addTask('prepare-venv', {
      description: `Create python virtual environment in ${taskOpts.venvPath}`,
      steps: [
        { exec: `${optsd.pythonExec} -m venv ${taskOpts.venvPath}` },
        {
          exec: `${taskOpts.venvPath}/bin/pip install pip-tools${PIP_TOOLS_VERSION}`,
        },
      ],
    });

    project.tasks.addTask('update-lockfile', {
      description: `Update lock file (${optsd.lockFile}) according to pyproject.toml`,
      steps: [
        {
          say: 'Prepare venv',
          spawn: 'prepare-venv',
        },
        {
          say: 'Updating lock file (runtime)',
          exec: `${taskOpts.venvPath}/bin/pip-compile --all-build-deps --output-file=${optsd.lockFile} pyproject.toml`,
        },
        {
          say: 'Updating lock file (dev)',
          exec: `${taskOpts.venvPath}/bin/pip-compile --all-build-deps --extra dev --strip-extras --output-file=${optsd.lockFileDev} pyproject.toml`,
        },
      ],
    });

    project.deps.addDependency(`pip-tools@${PIP_TOOLS_VERSION}`, DependencyType.DEVENV);

    if (taskOpts.attachTasksTo) {
      const attachTask = project.tasks.tryFind(taskOpts.attachTasksTo);
      if (!attachTask) {
        throw new Error(`'${taskOpts.attachTasksTo}' task not found`);
      }
      attachTask.spawn(prepareVenvTask);
      attachTask.spawn(installDevTask);
    }
  }

  public postSynthesize(): void {
    super.postSynthesize();

    this.project.logger.info('Initial preparation...');

    // run prepare venv, update lockfile and build
    const runtime = new TaskRuntime(this.project.outdir);
    runtime.runTask('prepare-venv');
    runtime.runTask('update-lockfile');
    runtime.runTask('build');
  }
}

export interface PipOptions {
  /**
   * Python executable path to be used while creating the virtual environment
   * used in this project
   * @default python
   */
  readonly pythonExec?: string;
  /**
   * Name of the file used to install dependencies from
   * This file is derived from pyproject.toml and have to be
   * manually updated if pyproject.toml is updated by using
   * the projen task 'update-lockfile'.
   * This lock won't include "dev" dependencies
   * Read more at https://stackoverflow.com/questions/34645821/pip-constraints-files
   * @default constraints.txt
   */
  readonly lockFile?: string;
  /**
   * Same as lockFile, but it includes all dev dependencies
   * @default constraints-dev.txt
   */
  readonly lockFileDev?: string;
}
