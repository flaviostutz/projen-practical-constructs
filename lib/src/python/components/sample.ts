/* eslint-disable no-new */
import { Component, Project, SampleDir } from 'projen';

/**
 * Python code sample.
 */
export class PythonBasicSample extends Component {
  constructor(project: Project) {
    super(project);

    // src directory

    // this sample will produce some formatting errors that are
    // used to exercise the use of lint-fix in the pipeline
    new SampleDir(project, 'src/hello_world', {
      files: {
        '__init__.py': '',
        'hello.py': `"""Says hello to the given name."""
def say(name: str) -> str:
              """Return a greeting message.
              Args:
                  name (str): Name to greet.
              Returns:
                  str: greeting message
              """
              return f"Hello {name}!"
          `,
        'py.typed': '',
      },
    });

    // tests directory
    new SampleDir(project, 'tests/hello_world', {
      files: {
        'test_hello_world.py': `from hello_world import hello

def test_hello_with_name():
    assert hello.say("Alice") == "Hello Alice!"

def test_hello_with_empty_string():
    assert hello.say("") == "Hello !"

def test_hello_with_special_characters():
    assert hello.say("!@#$") == "Hello !@#$!"
          `,
      },
    });
  }
}
