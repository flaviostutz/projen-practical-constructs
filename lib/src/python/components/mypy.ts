/* eslint-disable no-new */
import { Component, DependencyType, Project } from 'projen';

import { MyPyIniFile } from '../files';
import { addSpawnTaskToExisting, TaskOptions } from '../tasks';
import { CommonTargets } from '../../common/components/common-target-type';

export class MyPy extends Component {
  constructor(project: Project, taskOpts: TaskOptions) {
    super(project);

    new MyPyIniFile(project);
    project.deps.addDependency('mypy@1.14.*', DependencyType.DEVENV);
    project.addGitIgnore('.mypy_cache');
    const lintMypyTask = project.tasks.addTask('lint-mypy', {
      description: `Code type checks (mypy)`,
      exec: `${taskOpts.venvPath}/bin/mypy src tests`,
    });
    addSpawnTaskToExisting(project, lintMypyTask, CommonTargets.LINT);
  }
}
