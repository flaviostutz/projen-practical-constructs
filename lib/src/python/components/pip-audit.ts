/* eslint-disable no-new */
import { Component, DependencyType, Project } from 'projen';

import { addSpawnTaskToExisting, TaskOptions } from '../tasks';
import { CommonTargets } from '../../common/components/common-target-type';

export class PipAudit extends Component {
  constructor(project: Project, opts: TaskOptions) {
    super(project);

    project.deps.addDependency('pip-audit@2.*', DependencyType.DEVENV);
    project.addGitIgnore('.audit_cache');
    const lintAuditTask = project.tasks.addTask('lint-audit', {
      description: `Code type checks (mypy)`,
      exec: `${opts.venvPath}/bin/pip-audit --cache-dir .cache/.audit_cache`,
    });
    addSpawnTaskToExisting(project, lintAuditTask, CommonTargets.LINT);
  }
}
