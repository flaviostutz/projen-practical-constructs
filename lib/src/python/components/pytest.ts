/* eslint-disable no-new */
import { Component, DependencyType, Project } from 'projen';

import { addSpawnTaskToExisting, TaskOptions } from '../tasks';
import { PyTestIniFile, PyTestIniOptions } from '../files/pytest-ini';
import { CoveragercFile, CoveragercFileOptions } from '../files/coveragerc';
import { CommonTargets } from '../../common/components/common-target-type';

export class PyTest extends Component {
  constructor(project: Project, taskOpts: TaskOptions, opts?: PyTestOptions) {
    super(project);

    new PyTestIniFile(project);
    new CoveragercFile(project, opts);
    project.deps.addDependency('pytest@8.3.*', DependencyType.DEVENV);
    project.deps.addDependency('pytest-cov@6.0.*', DependencyType.DEVENV);
    project.addGitIgnore('.coverage');
    project.addGitIgnore('.pytest_cache');
    const lintUnitTask = project.tasks.addTask('test-unit', {
      description: `Unit tests (pytest)`,
      exec: `${taskOpts.venvPath}/bin/pytest`,
    });
    addSpawnTaskToExisting(project, lintUnitTask, CommonTargets.TEST);
  }
}

export interface PyTestOptions extends CoveragercFileOptions, PyTestIniOptions {}
