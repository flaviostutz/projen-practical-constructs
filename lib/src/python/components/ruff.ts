/* eslint-disable no-new */
import { Component, DependencyType, Project } from 'projen';

import { addSpawnTaskToExisting, TaskOptions } from '../tasks';
import { RuffTomlFile, RuffTomlFileOptions } from '../files/ruff-toml';
import { CommonTargets } from '../../common/components/common-target-type';

export class Ruff extends Component {
  constructor(project: Project, taskOpts: TaskOptions, opts?: RuffOptions) {
    super(project);

    new RuffTomlFile(project, opts);
    project.deps.addDependency('ruff@0.8.*', DependencyType.DEVENV);
    project.addGitIgnore('.ruff_cache');
    const lintRuffTask = project.tasks.addTask('lint-ruff', {
      description: `Code checks (RUFF)`,
      steps: [
        {
          exec: `${taskOpts.venvPath}/bin/ruff format --check src tests`,
        },
        {
          exec: `${taskOpts.venvPath}/bin/ruff check src tests`,
        },
      ],
    });
    addSpawnTaskToExisting(project, lintRuffTask, CommonTargets.LINT);

    const lintFixRuffTask = project.tasks.addTask('lint-ruff-fix', {
      description: `Lint fix (RUFF)`,
      steps: [
        {
          exec: `${taskOpts.venvPath}/bin/ruff format src tests`,
        },
        {
          exec: `${taskOpts.venvPath}/bin/ruff check --fix src tests`,
        },
      ],
    });
    addSpawnTaskToExisting(project, lintFixRuffTask, CommonTargets.LINT_FIX);
  }
}

export interface RuffOptions extends RuffTomlFileOptions {
  /**
   * Attach lint fix tasks to parent
   */
  readonly attachFixTaskTo?: string;
}
