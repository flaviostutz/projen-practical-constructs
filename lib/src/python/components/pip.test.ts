/* eslint-disable no-new */

import { Testing } from 'projen';

import { TestProject } from '../../common/test-project';
import { CommonTargets } from '../../common/components/common-target-type';

import { Pip } from './pip';

describe('Pip', () => {
  test('snapshot of default options', () => {
    const project = new TestProject();
    project.tasks.addTask(CommonTargets.INSTALL);
    new Pip(project, { venvPath: '.venv' });
    const out = Testing.synth(project);
    expect(out).toMatchSnapshot();
  });
  test('snapshot with custom options', () => {
    const project = new TestProject();
    project.tasks.addTask(CommonTargets.INSTALL);
    new Pip(project, { venvPath: '.custom_venv' }, { lockFile: 'custom_requirements.txt' });
    const out = Testing.synth(project);
    expect(out).toMatchSnapshot();
  });

  test('snapshot with another set of custom options', () => {
    const project = new TestProject();
    project.tasks.addTask(CommonTargets.INSTALL);
    new Pip(
      project,
      {
        venvPath: '.another_venv',
      },
      { lockFile: 'another_requirements.txt', lockFileDev: 'dev_requirements.txt' },
    );
    const out = Testing.synth(project);
    expect(out).toMatchSnapshot();
  });
});
