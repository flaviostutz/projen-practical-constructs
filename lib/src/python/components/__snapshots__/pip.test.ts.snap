// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pip snapshot of default options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
.venv/
*.egg-info
build
*.pyc
__pycache__
.cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "pip-tools",
        "type": "devenv",
        "version": "7.4.1",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "env": {
      "VENV_PATH": ".venv",
    },
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "install": {
        "description": "Install dependencies from constraints.txt",
        "name": "install",
        "steps": [
          {
            "exec": "$VENV_PATH/bin/pip install --require-virtualenv -c constraints.txt",
          },
        ],
      },
      "install-dev": {
        "description": "Install dependencies from constraints-dev.txt",
        "name": "install-dev",
        "steps": [
          {
            "exec": "$VENV_PATH/bin/pip install --require-virtualenv -c constraints-dev.txt --editable .[dev]",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "prepare-venv": {
        "description": "Create python virtual environment in .venv",
        "name": "prepare-venv",
        "steps": [
          {
            "exec": "python -m venv .venv",
          },
          {
            "exec": "$VENV_PATH/bin/pip install pip-tools==7.4.1 projen==0.91.6",
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
      "update-lockfile": {
        "description": "Update lock file (constraints.txt) according to pyproject.toml",
        "name": "update-lockfile",
        "steps": [
          {
            "say": "Prepare venv",
            "spawn": "prepare-venv",
          },
          {
            "exec": "$VENV_PATH/bin/pip-compile --all-build-deps --output-file=constraints.txt pyproject.toml",
            "say": "Updating lock file (runtime)",
          },
          {
            "exec": "$VENV_PATH/bin/pip-compile --all-build-deps --extra dev --strip-extras --output-file=constraints-dev.txt pyproject.toml",
            "say": "Updating lock file (dev)",
          },
        ],
      },
    },
  },
}
`;

exports[`Pip snapshot with another set of custom options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
.another_venv/
*.egg-info
build
*.pyc
__pycache__
.cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "pip-tools",
        "type": "devenv",
        "version": "7.4.1",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "env": {
      "VENV_PATH": ".another_venv",
    },
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "install": {
        "description": "Install dependencies from another_requirements.txt",
        "name": "install",
        "steps": [
          {
            "exec": "$VENV_PATH/bin/pip install --require-virtualenv -c another_requirements.txt",
          },
        ],
      },
      "install-dev": {
        "description": "Install dependencies from dev_requirements.txt",
        "name": "install-dev",
        "steps": [
          {
            "exec": "$VENV_PATH/bin/pip install --require-virtualenv -c dev_requirements.txt --editable .[dev]",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "prepare-venv": {
        "description": "Create python virtual environment in .another_venv",
        "name": "prepare-venv",
        "steps": [
          {
            "exec": "python -m venv .another_venv",
          },
          {
            "exec": "$VENV_PATH/bin/pip install pip-tools==7.4.1 projen==0.91.6",
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
      "update-lockfile": {
        "description": "Update lock file (another_requirements.txt) according to pyproject.toml",
        "name": "update-lockfile",
        "steps": [
          {
            "say": "Prepare venv",
            "spawn": "prepare-venv",
          },
          {
            "exec": "$VENV_PATH/bin/pip-compile --all-build-deps --output-file=another_requirements.txt pyproject.toml",
            "say": "Updating lock file (runtime)",
          },
          {
            "exec": "$VENV_PATH/bin/pip-compile --all-build-deps --extra dev --strip-extras --output-file=dev_requirements.txt pyproject.toml",
            "say": "Updating lock file (dev)",
          },
        ],
      },
    },
  },
}
`;

exports[`Pip snapshot with custom options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
.custom_venv/
*.egg-info
build
*.pyc
__pycache__
.cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "pip-tools",
        "type": "devenv",
        "version": "7.4.1",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "env": {
      "VENV_PATH": ".custom_venv",
    },
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "install": {
        "description": "Install dependencies from custom_requirements.txt",
        "name": "install",
        "steps": [
          {
            "exec": "$VENV_PATH/bin/pip install --require-virtualenv -c custom_requirements.txt",
          },
        ],
      },
      "install-dev": {
        "description": "Install dependencies from constraints-dev.txt",
        "name": "install-dev",
        "steps": [
          {
            "exec": "$VENV_PATH/bin/pip install --require-virtualenv -c constraints-dev.txt --editable .[dev]",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "prepare-venv": {
        "description": "Create python virtual environment in .custom_venv",
        "name": "prepare-venv",
        "steps": [
          {
            "exec": "python -m venv .custom_venv",
          },
          {
            "exec": "$VENV_PATH/bin/pip install pip-tools==7.4.1 projen==0.91.6",
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
      "update-lockfile": {
        "description": "Update lock file (custom_requirements.txt) according to pyproject.toml",
        "name": "update-lockfile",
        "steps": [
          {
            "say": "Prepare venv",
            "spawn": "prepare-venv",
          },
          {
            "exec": "$VENV_PATH/bin/pip-compile --all-build-deps --output-file=custom_requirements.txt pyproject.toml",
            "say": "Updating lock file (runtime)",
          },
          {
            "exec": "$VENV_PATH/bin/pip-compile --all-build-deps --extra dev --strip-extras --output-file=constraints-dev.txt pyproject.toml",
            "say": "Updating lock file (dev)",
          },
        ],
      },
    },
  },
}
`;
